File: mo5/magneto5-nano/src/serpackets.cpp
#include "serpackets.hpp"

bool send_prelude(uint32_t size32) {
	while(!Serial.availableForWrite());
	// type of the packet
	uint8_t type = SERPACKETS_PRELUDE;
	Serial.write((char*)&type, sizeof(uint8_t));
	// size of the file to be sent
	Serial.write((char*)&size32, sizeof(uint32_t));
	return true;
}

bool send_ready(void) {
	while(!Serial.availableForWrite());
	// type of the packet
	uint8_t type = SERPACKETS_READY;
	Serial.write((char*)&type, sizeof(uint8_t));
	return true;
}

bool send_ack(void) {
	while(!Serial.availableForWrite());
	// type of the packet
	uint8_t type = SERPACKETS_ACK;
	Serial.write((char*)&type, sizeof(uint8_t));
	return true;
}

bool send_end(void) {
	while(!Serial.availableForWrite());
	// type of the packet
	uint8_t type = SERPACKETS_END;
	Serial.write((char*)&type, sizeof(uint8_t));
	return true;
}

bool send_data(byte *buf, uint16_t size16) {
	while(!Serial.availableForWrite());
	// type of the packet
	uint8_t type = SERPACKETS_DATA;
	Serial.write((char*)&type, sizeof(uint8_t));
	// send the size of the packet
	if(size16 > SERPACKETS_MAX_SIZE) return false;
	Serial.write((char*)&size16, sizeof(uint16_t));
	// send the data forming the packet
	Serial.write(buf, size16);
	return true;
}



int32_t wait_prelude(void) {
	while(!Serial.available());
	// check the type of the packet received
	uint8_t recv_type;
	Serial.readBytes((char*)&recv_type, sizeof(uint8_t));
	if(recv_type != SERPACKETS_PRELUDE) return false;
	// read the size of the file to be received
	uint32_t size32;
	Serial.readBytes((char*)&size32, sizeof(uint32_t));
	return size32;
}

bool wait_ready(void) {
	while(!Serial.available());
	// check the type of the packet received
	uint8_t recv_type;
	Serial.readBytes((char*)&recv_type, sizeof(uint8_t));
	if(recv_type != SERPACKETS_READY) return false;
	return true;
}

bool wait_ack(void) {
	while(!Serial.available());
	// check the type of the packet received
	uint8_t recv_type;
	Serial.readBytes((char*)&recv_type, sizeof(uint8_t));
	if(recv_type != SERPACKETS_ACK) return false;
	return true;
}

int16_t wait_data(byte *buf) {
	while(!Serial.available());
	// check the type of the packet received
	uint8_t recv_type;
	Serial.readBytes((char*)&recv_type, sizeof(uint8_t));
	if(recv_type == SERPACKETS_END) return 0;
	if(recv_type != SERPACKETS_DATA) return -1;
	// read the size of the packet
	uint16_t size16;
	Serial.readBytes((char*)&size16, sizeof(uint16_t));
	if(size16 > SERPACKETS_MAX_SIZE) return -1;
	// next read the data forming the packet
	Serial.readBytes((char*)buf, size16);
	return size16;
}

File: mo5/magneto5-nano/serpackets.py
class SerPackets:
	
	MAX_SIZE = 320

	PRELUDE = b'\xFF'
	DATA = b'\xFE'
	END = b'\xFD'
	READY = b'\xFC'
	ACK = b'\xFB'

	ser = None
	
	def __init__(self, serobj):
		self.ser = serobj
	
	def close(self):
		self.ser.close()
	
	def send_prelude(self, size32):
		self.ser.write(self.PRELUDE)
		# 4 bytes encoding (uint32_t)
		fsize = size32.to_bytes(4, 'little')
		self.ser.write(fsize)
	
	def send_ready(self):
		self.ser.write(self.READY)
	
	def send_ack(self):
		self.ser.write(self.ACK)

	def send_end(self):
		self.ser.write(self.END)

	def send_data(self, buffer):
		#return self.__send_packet(self.DATA, buf, 0)
		self.ser.write(self.DATA)
		s = len(buffer)
		if s > self.MAX_SIZE:
			return False
		# 2 bytes encoding (uint16_t)
		pktsize = s.to_bytes(2, 'little')
		self.ser.write(pktsize)
		self.ser.write(buffer)
		
	def wait_prelude(self):
		recv_type = self.ser.read(1)
		if recv_type != self.PRELUDE:
			return 0
		else:
			# 4 bytes encoding (uint32_t)
			fsize = int.from_bytes(self.ser.read(4), 'little')
			return fsize

	def wait_ready(self):
		recv_type = self.ser.read(1)
		return recv_type == self.READY

	def wait_ack(self):
		recv_type = self.ser.read(1)
		return recv_type == self.ACK
	
	def wait_data(self):
		recv_type = self.ser.read(1)
		if recv_type != self.DATA:
			if recv_type == self.END:
				return b''
			else:
				return None
		# 2 bytes encoding (uint16_t)
		pktsize = int.from_bytes(self.ser.read(2), 'little')
		buffer = self.ser.read(pktsize)
		return buffer
